datadog-ffe/src/rules_based/eval/eval_assignment.rs

```diff
@@ -0,0 +1,300 @@
+pub(super) enum AllocationNonMatchReason {
+    BeforeStartDate,
+    AfterEndDate,
+    FailingRule,
+    TrafficExposureMiss,
+}
+
+impl Allocation {
+    fn get_matching_split(
+        &self,
+        subject: &EvaluationContext,
+        now: Timestamp,
+    ) -> Result<(&Split, AssignmentReason), AllocationNonMatchReason> {
+        if self.start_at.is_some_and(|t| now < t) {
+            return Err(AllocationNonMatchReason::BeforeStartDate);
+        }
+        if self.end_at.is_some_and(|t| now > t) {
+            return Err(AllocationNonMatchReason::AfterEndDate);
+        }
+
+        let is_allowed_by_rules =
+            self.rules.is_empty() || self.rules.iter().any(|rule| rule.eval(subject));
+        if !is_allowed_by_rules {
+            return Err(AllocationNonMatchReason::FailingRule);
+        }
+
+        let split = self
+            .splits
+            .iter()
+            .find(|split| {
+                let matches = split.matches(subject.targeting_key());
+                matches
+            })
+            .ok_or(AllocationNonMatchReason::TrafficExposureMiss)?;
+
+        // Determine the reason for assignment
+        let reason = if !self.rules.is_empty() || self.start_at.is_some() || self.end_at.is_some() {
+            AssignmentReason::TargetingMatch
+        } else if self.splits.len() == 1 && self.splits[0].shards.is_empty() {
+            AssignmentReason::Static
+        } else {
+            AssignmentReason::Split
+        };
+
+        Ok((split, reason))
+    }
+}
+
+impl Split {
+    /// Return `true` if `targeting_key` matches the given split.
+    ///
+    /// To match a split, subject must match all underlying shards.
+    fn matches(&self, targeting_key: &str) -> bool {
+        self.shards.iter().all(|shard| shard.matches(targeting_key))
+    }
+}
+
+impl Shard {
+    /// Return `true` if `targeting_key` matches the given shard.
+    fn matches(&self, targeting_key: &str) -> bool {
+        let h = self.sharder.shard(&[targeting_key]);
+
+        self.ranges.iter().any(|range| range.contains(h))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::HashMap,
+        fs::{self, File},
+        sync::Arc,
+    };
+
+    use chrono::Utc;
+    use serde::{Deserialize, Serialize};
+
+    use crate::rules_based::{
+        eval::get_assignment,
+        ufc::{AssignmentValue, UniversalFlagConfig, VariationType},
+        Attribute, Configuration, EvaluationContext, Str,
+    };
+
+    #[derive(Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    struct TestCase {
+        flag: String,
+        variation_type: VariationType,
+        default_value: serde_json::Value,
+        targeting_key: Str,
+        attributes: Arc<HashMap<Str, Attribute>>,
+        result: TestResult,
+    }
+
+    #[derive(Debug, Serialize, Deserialize)]
+    struct TestResult {
+        value: serde_json::Value,
+    }
+
+    #[test]
+    fn evaluation_sdk_test_data() {
```

Comment:
It feels a bit weird that all the test cases are run inside a single method. I think it would be better to use a `build.rs` file to generate them. I'll open a small PR.